### ISSUE TITLE
Store c++ reference to temporary

### ISSUE THREAD

**Issue by mdavidsaver** [cython/cython#3759]
**Describe the bug**
I have a c++ API which uses the builder pattern (using a c++ object in lieu of named arguments).
I find that in some cases a reference to the builder object is stored, while the builder itself is not.
This results in a dangling stack pointer.

**To Reproduce**
Code to reproduce the behaviour:
```cython
# distutils: language = c++
#cython: language_level=3

cdef extern from *:
    cdef cppclass Builder:
        Builder& modify() except+
        void exec() except+

    cdef cppclass Context:
        @staticmethod
        Builder start()

cdef class X:
    def doit(self):
        Context.start().modify().exec()
```

The relevant portions from the generated c++ code for the `doit()` method.

```c++
  __Pyx_FakeReference<Builder> __pyx_t_1;
  ...
 /* "test.pyx":15
 * cdef class X:
 *     def doit(self):
 *         Context.start().modify().exec()             # <<<<<<<<<<<<<<
 */
  try {
    __pyx_t_1 = Context::start().modify();
```

The `start()` method returns a temporary `Builder` instance, and the `modify()` method
is returning a reference to this same `Builder`.  However, the `Builder` goes out of scope after this line
and `__pyx_t_1` is left with a dangling stack pointer.

```c++
  } catch(...) {
    __Pyx_CppExn2PyErr();
    __PYX_ERR(0, 15, __pyx_L1_error)
  }
  try {
    __pyx_t_1->exec();
  } catch(...) {
    __Pyx_CppExn2PyErr();
    __PYX_ERR(0, 15, __pyx_L1_error)
  }
...
```

**Expected behavior**

I was expecting that either the temporary `Builder` would be explicitly stored to keep it alive,
or that the line would not be split up between two try/catch blocks.

**Environment (please complete the following information):**
 - OS: Linux
 - Version: verified on 0.29.2 and 0.29.21

**Additional context**

I've found two workarounds.  Marking the `start()` method with `except+`,
or explicitly assigning a local variable, cause the `Builder` to be stored
and the reference to remain valid.

```cython
        Builder start() except+
```

or

```cython
    def doit(self):
        cdef Builder temp = Context.start()
        temp.modify().exec()
```

I got luck and (after sifting through the false positives) valgrind picked this right up.

**Comment by da-woods** [comment]
> I was expecting that either the temporary Builder would be explicitly stored to keep it alive,

The trouble is that Cython doesn't know that this temporary builder is same one you return a reference to. It isn't obvious that it must keep this alive.

>  [...] or that the line would not be split up between two try/catch blocks.

That would sacrifice some detail in the tracebacks, and also stop you from using [some of the more advanced exception handling on a per-function basis](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions).

--------

I think your second (explicit) workaround is better, just because it's explicit. Cython will re-use temporaries once it thinks they're finished with which may cause you problems with the first version occasionally.

----------

Unfortunately I doubt that this is fixable without breaking too much existing code. It unfortunately doesn't look like a pattern that interacts well with Cython.

**Comment by mdavidsaver** [comment]
> Cython doesn't know that this temporary builder is same one you return a reference to

I agree that Cython does not have enough information to infer this.  The only reasonable way I can see to accomplish this would be to emulate the lifetime of temporaries as a c++ compiler would.  I think this would involve a more complex notion of local variable scope than Cython seems to have at present (based on my reading of generated code).  So I suspect this would be a major change.

(This isn't a criticism.  Simple is good, but can be at odds with c++)


> Cython will re-use temporaries once it thinks they're finished

I noticed and wondered about this.  Is this intentional, or a side effect?  (eg. of not using [SSA](https://en.wikipedia.org/wiki/Static_single_assignment_form))  As I've found cython so useful, it has been on my mind to dig into it.  With such a large project though, it is difficult to know where to start.

**Comment by da-woods** [comment]
> I agree that Cython does not have enough information to infer this. The only reasonable way I can see to accomplish this would be to emulate the lifetime of temporaries as a c++ compiler would

This turns out to be surprisingly hard. Cython will generate gotos that jump across variable definitions (which C++ doesn't like), hence everything really needs to be defined at the start of the function. It's really defined with C in mind so the C++ code generated isn't always great.

> I noticed and wondered about this. Is this intentional, or a side effect? (eg. of not using SSA)

It's intentional, presumably with a view to minimizing the stack size. I'm not sure how much practical difference it makes though.

> As I've found cython so useful, it has been on my mind to dig into it. With such a large project though, it is difficult to know where to start.

The place to start if you're genuinely interested is look at a simple-looking bug (making temporary variables behave in a C++-like manner is _not_ a simple bug though!) There's a [semi-useful wiki page](https://github.com/cython/cython/wiki/HackerGuide) about how things are structured.
